toolchain/check/eval.cpp

```diff
@@ -1311,6 +1324,43 @@ static auto MakeConstantForBuiltinCall(Context& context, SemIRLoc loc,
       return SemIR::ConstantId::NotConstant;
     }
 
+    case SemIR::BuiltinFunctionKind::TypeAnd: {
+      CARBON_CHECK(arg_ids.size() == 2);
+      auto lhs_facet_type_id = SemIR::FacetTypeId::None;
+      auto rhs_facet_type_id = SemIR::FacetTypeId::None;
+      for (auto [facet_type_id, arg_id] :
+           llvm::zip(std::to_array({&lhs_facet_type_id, &rhs_facet_type_id}),
+                     arg_ids)) {
+        if (auto facet_type =
+                context.insts().TryGetAs<SemIR::FacetType>(arg_id)) {
+          *facet_type_id = facet_type->facet_type_id;
+        } else {
+          CARBON_DIAGNOSTIC(
+              FacetTypeRequiredForTypeAndOperator, Error,
+              "BitAnd operator on types requires facet type on both sides");
```

Comment:
How about something like:

```suggestion
              "{0} is not a facet type", SemIR::TypeId);
```

(passing `context.types().GetTypeIdForTypeInstId(arg_id)`)

As I think more about the repeated diagnostic -- https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/docs/diagnostics.md#diagnostic-message-style-guide says "Diagnostics should describe the situation the toolchain observed. The language rule violated can be mentioned if it wouldn't otherwise be clear."

The current phrasing sounds like a language rule, maybe the suggested message is closer to the source of the error? You could add something like "; using `&` on types requires facet types` if you think it's needed.

Reply:
Right true, it's "a non-facet type was found here" error. The downside is that it has to be phrased in the negative with "non-" there. I don't know what name to use for a type that isn't a facet type otherwise.

Reply:
This SG, though why aren't you including the type ID?

Note abstractly:

```
fn GetC() -> type { return C; }
fn GetA() -> type { return A; }

// Should be happy.
({} as C) as (C as (GetA() & GetA()));
// Would be helpful to print `C` here.
({} as C) as (C as (GetC() & GetC()));
```

Reply:
Ah that's a good idea, I hadn't thought of it.
```
+  // CHECK:STDERR: fail_combine_with_non_facet_type.carbon:[[@LINE+4]]:23: error: non-facet type ``C`` combined with `&` operator [FacetTypeRequiredForTypeAndOperator]
```
That's better :) Thanks!